### PR TITLE
Added support for core-metadata test suite for non-docker environment…

### PR DESCRIPTION
**Pull request by bhavikkothari-eaton**
This PR is regarding below changes

1. Added below scripts under "TAF/utils/scripts/manual/" to execute EdgeX-TAF under non-docker environment.
    - deploy-edgex.sh: This script is to deploy KVM and get IP address
    - api-gateway-token.sh: This script is to generate X-Auth-Token
    - shutdown.sh: This script is to shutdown edgex after test execution
    - restart-service.sh: This script is to restart any service
    - stop-service.sh: This script is to stop any service
2. Added trigger,.sh script. This is our main script which will execute in our pipeline.
3. Added prerequisite_changes.py file. This file contains changes to be done before execution.
4. Added README_Instruction.md file which will help to understand how execution flow will be.

### CODE REVIEW

File: README_Instructions.md
Comment (markabrezovec-eaton) [inline README_Instructions.md]: Need to be more specific than 'socket port'.  This is the port for the EdgeX client, right?

Comment (markabrezovec-eaton) [inline README_Instructions.md]: Right now you only specify 8002 (https).  Does it not work with 8001 (http)?  If that is the case, we should give a little more information on why.

File: README_Instructions.md
Comment (markabrezovec-eaton) [inline README_Instructions.md]: Calling this input 'module' doesn't quite feel right.  Is something like 'dut' (for Device Under Test) or 'target' more appropriate?

@rajatkumar-eaton  - I am interested in your thoughts.

File: README_Instructions.md
Comment (markabrezovec-eaton) [inline README_Instructions.md]: We use data.txt throughout these files.  I think one script populates that data and shares it (or something like that).  Could we add a comment for that somewhere in this MD file describing that workflow?

File: TAF/utils/scripts/manual/deploy-edgex.sh
Comment (markabrezovec-eaton) [inline TAF/utils/scripts/manual/deploy-edgex.sh]: From what I can tell, we are assuming that if this file exists, the KVM is running.  You delete this file between runs, so that helps.  That said, it is not a guarantee that the KVM is running, especially if a person runs this test locally.

Is there a way to check that the KVM is actually running when it finds the file?  For instance, pull out the PID and check the server to see if that PID is still actively running?

File: TAF/utils/scripts/manual/shutdown.sh
Comment (markabrezovec-eaton) [inline TAF/utils/scripts/manual/shutdown.sh:12]: I am guessing we never figured out why "kill -2" does not work when run from this script?  Maybe timebox the next investigation to an hour.  Make sure to:

A) Make sure the command is correct.
B) Make sure the output is verbose.
C) Check the output text from the command and see if it gives you any guidance on what happened.
D) Check the return code from the command and see what it returns

File: prerequisite_changes.py
Comment (markabrezovec-eaton) [inline prerequisite_changes.py]: Could we get a comment on each of these changes on why the change is needed?  A short, single sentence should help.  This will be important for the future if EdgeX-TAF changes anything.

File: trigger.sh
Comment (markabrezovec-eaton) [inline trigger.sh]: Should we have a command to upgrade pip earlier in the file?

File: trigger.sh
Comment (markabrezovec-eaton) [inline trigger.sh:123]: Shouldn't we be using the requirements.txt file provided by this repo instead of doing each one individually?

If that file somehow does not include all the libraries we need, we should create a second requirements file (requirements_eaton.txt) to cover the remainder.  That said, I don't see us adding anything special of our own, so I would not anticipate needing anything different.  Installing this requirements.txt and the one from edgex-taf-common should be sufficient unless I am missing something.

File: README_Instructions.md
Comment (bhavikkothari-eaton) [reply README_Instructions.md]: Yes, correct

File: README_Instructions.md
Comment (bhavikkothari-eaton) [reply README_Instructions.md]: Okay

File: README_Instructions.md
Comment (bhavikkothari-eaton) [reply README_Instructions.md]: Mark it is working with 8001 port but as of now we are going with 8002 port.

File: README_Instructions.md
Comment (bhavikkothari-eaton) [reply README_Instructions.md]: @markabrezovec-eaton I will put name as "target"

File: TAF/utils/scripts/manual/deploy-edgex.sh
Comment (bhavikkothari-eaton) [reply TAF/utils/scripts/manual/deploy-edgex.sh]: yes , using below command we can check that "start_kvm.py" process is running or not.

"**ps ax | grep -v grep | grep start_kvm.py**"

So, the flow will be like if "kvm_data.json" file is there then will check that "start_kvm.py" process is running or not. and if process is running then will not deploy edgex again.

File: prerequisite_changes.py
Comment (bhavikkothari-eaton) [reply prerequisite_changes.py]: ok sure

File: trigger.sh
Comment (bhavikkothari-eaton) [reply trigger.sh:123]: ok. I will create one separate requirements_eaton.txt file which includes robot framework related libraries. so there will be 2 requirements.txt files. one is our requiremets_eaton.txt and another one is for edgex-taf-common.

File: trigger.sh
Comment (rajatkumar-eaton) [inline trigger.sh:1]: @bhavikkothari-eaton Just a query:
Whatever services or pre-requisite we are installing in this file, would that be bypass. if already present those setup locally or going to overwrite the same?

File: trigger.sh
Comment (rajatkumar-eaton) [reply trigger.sh]: @bhavikkothari-eaton Shouldn't we think of keeping it uniform here also as I saw many places below section of the file used as "pip3" so we can direct use the same here.
Thought?

File: trigger.sh
Comment (bhavikkothari-eaton) [reply trigger.sh:1]: @rajatkumar-eaton as of now it will overwrite the same. because we are assuming that every time in pipeline we are cloning new edgex-taf. so changes will not be there

File: trigger.sh
Comment (bhavikkothari-eaton) [reply trigger.sh]: @markabrezovec-eaton after activating venv "python3 -m pip install -U pip" command will upgrade pip version in virtual environment

@rajatkumar-eaton I will change "pip install -U virtualenv" to pip3

File: TAF/utils/scripts/manual/shutdown.sh
Comment (bhavikkothari-eaton) [reply TAF/utils/scripts/manual/shutdown.sh:12]: @markabrezovec-eaton I have checked return status of "kill -2" command. but it is 0 so it not showing any error or executing command. and I also checked output of command but I did not find any guidance.

File: trigger.sh
Comment (markabrezovec-eaton) [reply trigger.sh:123]: To clarify, I am still not convinced we need a requirements_eaton.txt file.  That is a strategy we should use only if we need custom libraries for our own scripts.  That said, the only libs I see you importing in your python script are 'sys' and 'fileinput', which should both be built-in.  Per my understanding, we should be able to run EdgeX-TAF installing only the libraries provided by their requirements files, i.e. the one in this repo and the on from edgex-taf-common.

File: trigger.sh
Comment (markabrezovec-eaton) [inline trigger.sh]: This call needs updated.  You are correct that the command is "python3 -m venv <directory>", but the recommendation is to give it a name.  So your command should not set just directory to '.'.  It should instead be something like:

    python3 -m venv env_taf

File: trigger.sh
Comment (markabrezovec-eaton) [inline trigger.sh]: This line is actually not needed.  The 'virtualenv' library is only for versions lower than Python 3.4

File: trigger.sh
Comment (markabrezovec-eaton) [inline trigger.sh]: Building on above, I think you only need the following:

    echo "<<<<< Installing Robot framework related libraries >>>>>"
    pip3 install -r requirements.txt
    sleep 1

    # Install TAF common:
    echo "<<<<< Installing TAF common >>>>>"
    git clone https://github.com/edgexfoundry/edgex-taf-common.git
    sleep 1
    
    # Install dependency lib
    echo "<<<<< Installing TAF common related libraries >>>>>"
    pip3 install -r ./edgex-taf-common/requirements.txt
    
    # Install edgex-taf-common as lib
    pip3 install ./edgex-taf-common
    ...

If you get any errors running EdgeX-TAF without the requirements_eaton.txt, let me know.  I would like to see what those errors are.

File: trigger.sh
Comment (markabrezovec-eaton) [inline trigger.sh:122]: I see a problem.  See line 117.  You activate the VENV there.  So your scripts are only actually in the VENV if you make it as a part of this script.

Two possible workflows:
1.  If it finds no VENV, it will create one and activate it,
2. If it finds the VENV has already been created, it will not 'activate'.  That means you won't actually use the VENV for the rest of this bash script.

For workflow 2, you will need to have an 'else' clause to activate.

File: requirements_eaton.txt
Comment (markabrezovec-eaton) [inline requirements_eaton.txt]: Per below, I don't actually think we will need this if we clean things up and run it properly.  Let me know if you continue to find issues.

File: README_Instructions.md
Comment (markabrezovec-eaton) [reply README_Instructions.md]: Marking this 'Unresolved' until the change gets merged to PR.  (Just being a bit picky.)

File: TAF/utils/scripts/manual/shutdown.sh
Comment (markabrezovec-eaton) [reply TAF/utils/scripts/manual/shutdown.sh:12]: If only kill -9 works, go ahead and use it.  Add a comment indicating that we would prefer kill -2 in the future.

**Review by markabrezovec-eaton — COMMENTED** [review]
I am not sure why all of my comments were changed to 'Pending', but I think I need to add this top level comment for my stuff to be visible.

File: trigger.sh
Comment (bhavikkothari-eaton) [reply trigger.sh]: sure

File: README_Instructions.md
Comment (rajatkumar-eaton) [reply README_Instructions.md]: > Calling this input 'module' doesn't quite feel right. Is something like 'dut' (for Device Under Test) or 'target' more appropriate?
> 
> @rajatkumar-eaton - I am interested in your thoughts.

Yes @markabrezovec-eaton I completely agree on this. We should use DUT or Target Device or simply Target.

File: TAF/utils/scripts/manual/shutdown.sh
Comment (bhavikkothari-eaton) [reply TAF/utils/scripts/manual/shutdown.sh:12]: ok

File: README_Instructions.md
Comment (bhavikkothari-eaton) [reply README_Instructions.md]: Ok, I will make it as a Target.

File: trigger.sh
Comment (bhavikkothari-eaton) [reply trigger.sh:123]: I tested locally by removing requirements_eaton.txt file (also uninstalled same libraries from my local system) and simply installing requirements.txt from EdgeX-TAF and edgex-taf-common repo and it works. so will remove requirements_eaton.txt file.

File: trigger.sh
Comment (bhavikkothari-eaton) [reply trigger.sh]: I made changes as above and did not find any error.

File: trigger.sh
Comment (bhavikkothari-eaton) [reply trigger.sh]: ok will remove this line.

File: requirements_eaton.txt
Comment (bhavikkothari-eaton) [reply requirements_eaton.txt]: I totally remove this file. So, while executing I did not find any issues after removing this

File: trigger.sh
Comment (bhavikkothari-eaton) [reply trigger.sh:122]: Ok, I will add else part to activate it.

File: trigger.sh
Comment (markabrezovec-eaton) [inline trigger.sh]: If the env is created, we can assume pip is already installed.  We can probably just remove this one line.

**Comment by markabrezovec-eaton** [comment]
I made one small comment.  Otherwise this has my tentative approval.  @rajatkumar-eaton could you provide the final review?

After merging, I recommend we demonstrate this to Anil and development (at least Pune side) and get some preliminary feedback.  For now we can call it a basic framework they can use in their bench testing.  A pipeline for this is not yet planned until PI02, at the earliest.

File: trigger.sh
Comment (bhavikkothari-eaton) [reply trigger.sh]: ok will remove this line.

File: README_Instructions.md
Comment (rajatkumar-eaton) [inline README_Instructions.md]: @bhavikkothari-eaton This readme file seems to appear as created to run only in pipeline, can we put generic term or to include "Pipeline as well as in local environment" ?
Thought?

File: TAF/utils/scripts/manual/api-gateway-token.sh
Comment (rajatkumar-eaton) [inline TAF/utils/scripts/manual/api-gateway-token.sh]: @bhavikkothari-eaton We are not opening swagger UI page, just suing through curl command through command line.
Should we still use "Swagger login" in comment?

File: README_Instructions.md
Comment (rajatkumar-eaton) [inline README_Instructions.md:30]: @bhavikkothari-eaton I see only explained about "trigger.sh" and "prerequisite_changes.py" in readme file here.
Shouldn't we add few lines about other .sh file as well like what it does, how and why to run them?

Thought?

File: README_Instructions.md
Comment (bhavikkothari-eaton) [reply README_Instructions.md]: @rajatkumar-eaton Actually I made this script mainly to run in pipeline because while deploying edgex we are using "**--yocto --conn_method=http**" which is suitable for pipeline only. for local environment we need to remove this line. 

I can mention in this as a note in readme file and can make it pipeline as well as local environment

File: TAF/utils/scripts/manual/api-gateway-token.sh
Comment (bhavikkothari-eaton) [reply TAF/utils/scripts/manual/api-gateway-token.sh]: We are not using "Swagger login" directly. I will change it as "This script will generate X-Auth-Token".

File: README_Instructions.md
Comment (bhavikkothari-eaton) [reply README_Instructions.md:30]: We can mention other scripts under manual directory. Also will mention workflow of these scripts.

**Review by rajatkumar-eaton — APPROVED** [review]

**Pull request merged**
